---
title: "Wilkens-Bishop Dec 2021 - Revisiting PacBio pipeline results in Oct 2023"
author: "Jessica Holmes, Chris Fields"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    pdf_print: paged
    fig_height: 6
    fig_width: 12
    toc: yes
    toc_float: yes
  word_document:
    toc: yes
    fig_height: 6
    fig_width: 12
  pdf_document:
    toc: yes
  powerpoint_presentation:
    toc: no
    fig_height: 6
    fig_width: 12
---

# Intro

This covers agglomeration steps and prevalence filtering for the phthalate exposure in mice study.

# Set up

Code (not shown in the report) is initialized and loaded here.  We don't include the code in the report but make this available as needed; please see the [Github repository](https://github.com/HPCBio/flaws-2020March-16S)for this project for the final version.

```{r PrevalenceFiltering-1, echo=TRUE, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

```{r PrevalenceFiltering-2, include=FALSE}
# Note that not all libraries will be needed.  Most phyloseq code uses ggplot and tidyverse internally, therefore we explicitly load here
library(knitr)
library(tidyverse)
library(phyloseq)

# this seems to have issues with caching and phyloseq
# library(ggtree) 

# For normalization
library(metagenomeSeq)

# phylogenetic tree input
library(ape)

# read/modify BIOM 
library(biomformat)

# ggplot functions for trees and dendrograms
library(ggdendro)

# distance measures, PERMANOVA, ANOSIM
library(vegan)

# generation of stats values for graphs
library(ggpubr)

# normalization (CLR)
library(mixOmics)

# to get labels2color
library(WGCNA)

# mixed models (needs to be updated)
library(lme4)
library(lmerTest)
library(nlme)

# sample decontamination
#library(decontam)

# to get post-hoc tests for mixed-model tests 
library(lsmeans)
library(devtools)

#Other
library(gridExtra)

# needed in case we want to use ANCOM
#library(exactRankTests)

# this is to load some extension helper code, see: https://github.com/HPCBio/phyloseq-extended
devtools::load_all('~/src/phyloseq-extended')
```

```{r PrevalenceFiltering-3, include=FALSE}
# Setting up the analysis, including adding helper functions.  The document won't include the actual code, but the functions are present in the Rmd document.  The functions here include ones to:
options(stringsAsFactors = FALSE)
theme_set(theme_bw())
```

```{r PrevalenceFiltering-4, include=FALSE}
# Remove the tags on the taxonomic ranks, which are redundant with the column headers.
stripTaxaTags <- function(physeq) {
  oldMA <- as(tax_table(physeq), "matrix")
  newMA <- apply(oldMA, 2, function(x) {sub('\\w__','', x)})
  if (inherits(physeq, "taxonomyTable")) {
      return(tax_table(newMA))
  }
  else {
      tax_table(physeq) <- tax_table(newMA)
      return(physeq)
  }
}
```

```{r PrevalenceFiltering-5 }
# Convert sequences to names (culled from https://github.com/LangilleLab/microbiome_helper/blob/master/convert_dada2_out.R) 

renameTaxIds <- function(physeq, file.name="seqs.fasta") {
  suppressMessages(require("ShortRead"))
  seqtab.physeq <- otu_table(physeq)
  seqs <- colnames(seqtab.physeq)
  ids_study <- paste("seq", 1:ncol(seqtab.physeq), sep = "_")
  seqs.dna <- ShortRead(sread = DNAStringSet(seqs), id = BStringSet(ids_study))
  # Write out fasta file.
  writeFasta(seqs.dna, file = file.name)
  taxa_names(physeq) <- ids_study
  # TODO: add the sequences back to the phyloseq instance
  # physeq <- merge_phyloseq(physeq)
  return(physeq)
}
```

```{r PrevalenceFiltering-6}
# original code: https://github.com/twbattaglia/btools/blob/master/R/estimate_pd.R
estimate_pd <- function(phylo) {
  # Error if input is not of class phylo
  if(class(phylo) != "phyloseq"){
    stop("Input file is not of class 'phyloseq'.")
  }

  # Error if no class phy_tree
  if(!(.hasSlot(phylo, "phy_tree"))){
    stop("Could not find tree slot in phylo object.")
  }
  
  if (!require('picante')) stop("Function requires the picante library.")

  # Transpose if needed
  # Adapted from phyloseq/vegan import
  OTU <- phyloseq::otu_table(phylo)
  if (taxa_are_rows(OTU)) {
    OTU <- t(OTU)
  }

  # Get matrix version of OTU table
  otutable <- as(OTU, "matrix")

  # Get phylogenetic tree from phyloseq object
  tree <- phyloseq::phy_tree(phylo)

  # Print status message
  message("Calculating Faiths PD-index...")

  # If object is greater than 10mb, then print status message
  if(object.size(otutable) > 10000000){
    message("This is a large object, it may take awhile...")
  }

  # Calculate Faith's PD-index
  #
  pdtable <- picante::pd(otutable, tree, include.root = F)

  # Return data frame of results
  return(pdtable)
}
```

```{r PrevalenceFiltering-7}
# CLR normalization 
# (from McMurdie (Meth Mol Bio 2018) supplemental package)
zero_comp = function(x){
  if(taxa_are_rows(x)){x <- t(x)}
  matx = otu_table(x)
  # `zCompositions::cmultRepl` expects the samples to be in rows and OTUs to be in columns
  matxzc = zCompositions::cmultRepl(matx, method="CZM", output="p-counts")
  otu_table(x) <- otu_table(matxzc, taxa_are_rows = FALSE)
  return(x)
}
# CLR definition
geometric_mean = function(x){
  exp(mean(log(x)))
}
clr = function(x, base=2){
  x <- log((x / geometric_mean(x)), base)
}
phyloseq_CLR = function(physeq){
  suppressMessages({physeq <- zero_comp(physeq)})
  return(transform_sample_counts(physeq, fun = clr))
}
```

# Load data

Load in the filtered data from part 1:

```{r AlphaDiversity-PacBio-8}
physeq.filtered <- readRDS('./results/phyloseq.filtered.pt1.RDS')
physeq.filtered
```

# Additional Filtering

We performed some high level filtering to remove artifacts and problematic data. Next step is agglomeration of count data and prevalence filtering.

## Explore taxon data 

What is the range in total counts per taxon?

```{r PrevalenceFiltering-8 }
range(taxa_sums(physeq.filtered))
```

No taxa have less than 2 counts

What does the distribution look like at the low end?
```{r PrevalenceFiltering-9 }
hist(log2(taxa_sums(physeq.filtered)), 1000)
```
PacBio: There's a lot on the low end, that will likely need to be filtered out.
Shoreline: Not a whole lot at the low end and looks like a normal distribution.

What about sample counts?  What is the range in total counts per sample?
```{r PrevalenceFiltering-10 }
range(sample_sums(physeq.filtered))
```

We have some on the low end, with `r sum(sample_sums(physeq.filtered) <= 5000)` samples less than 5k counts.

```{r PrevalenceFiltering-11 }
p <- ggplot(data = data.frame(
    SampleSums = sample_sums(physeq.filtered),
    Names = factor(sample_names(physeq.filtered), ordered = TRUE,
                   levels = sample_names(physeq.filtered)),
    Group = factor(sample_data(physeq.filtered)$Location, ordered = TRUE)
), aes(y = SampleSums, x = Names, fill = Group))
p <- p + geom_bar(stat = 'identity' )
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p
```

One issue is the library sizes vary considerably, and there's no real pattern to it.

How do the ASV counts correlate with the read counts?

```{r PrevalenceFiltering-12 }
myData <- data.frame(
  Name = sample_names(physeq.filtered),
  OTUSums = sample_sums(physeq.filtered),
  Reads = as.numeric(sample_data(physeq.filtered)$input),
  Group = sample_data(physeq.filtered)$Location
)
p <- ggplot(data = myData, aes(x = Reads, y = OTUSums))
p <- p + geom_smooth(method = "gam", color = "lightgreen")
p <- p + geom_smooth(method = "lm", color = "lightblue")
p <- p + geom_point(aes(color = Group))
p
```

Seems to be quite linear.

Next we filter based on the features prevalent in the samples.  We will also switch the order of the filtering and tree-based (tip) agglomeration. If the tips are noisy (strain-level variation that is difficult to assign), then we'll proceed with this type of filtering.

## Tip agglomeration

What does the current tree look like?

```{r PrevalenceFiltering-13}
p <- plot_tree(physeq.filtered, 
          nodelabf = nodeplotblank, 
          color="Sample", 
          ladderize = "left", 
          method = "treeonly") +
  ggtitle(paste0("Original tree: ", ntaxa(physeq.filtered), " taxa")) +
  theme(plot.title = element_text(size = 10))
library(plotly)

ggplotly(p)
```

Zooming into the tips indicates there are a many sequences with very small differences, so tip agglomeration should be useful in this case.

What is the distribution of tree edge lengths?

```{r PrevalenceFiltering-14}
hist(log(phy_tree(physeq.filtered)$edge.length), 
     xlab = "Edge Length (log)", 
     main = "Edge length distribution")
```
PacBio: There are a lot on the short end, and none on the long end (beyond zero). pretty clear split between short and long.
Shoreline: There are a small number on the short end, and even less on the long end.


### Clip out long branches

```{r PrevalenceFiltering-14.B}
tmp <- phy_tree(physeq.filtered)

# grab the tip lengths and format for ggplot
# note the tip length is actually log scale to view this at a reasonable scale
treeTips <- data.frame(
  ID = tmp$tip.label,
  Tip.Length = log(tmp$edge.length[tmp$edge[,2] <= Ntip(tmp)])
)

p <- treeTips %>%
  ggplot( aes(x=Tip.Length, fill = "black")) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', bins = 100)

ggplotly(p)
```

You have to zoom in to the right to visualize any long branchs:

```{r}
p + xlim(-5, 1)
```

Nothing stands out as being too long. Here are the top 5 lengths to be sure.

```{r}
longbranch <- treeTips[order(treeTips$Tip.Length, decreasing = TRUE)[1:5],]
knitr::kable(longbranch)
```

What classifications are at the top of this list?

```{r}
tmp2 <- cbind(tax_table(physeq.filtered), as.data.frame(taxa_sums(physeq.filtered)))

knitr::kable(tmp2[longbranch$ID,])
```

PacBio: The top 5 are to the Family or Order level  and don't seem too odd. Can probably leave them. 
Shoreline: The top 5 are only classified up to the Phylum or Class level, so it wouldn't hurt to remove them. I also ran the following steps without removing them, and the results ended up being the same after prevalence filtering. 

```{r}
#What samples are these in?
tmp <-suppressWarnings(prune_taxa(taxa_names(physeq.filtered) %in% longbranch$ID,
                  physeq.filtered))

ssums <- sample_sums(tmp)
ssums[ssums > 0]
```

PacBio: It appears that these are all only present in one sample, so they'll likely get filtered downstream anyway.

```{r}
# physeq.filtered <- prune_taxa(!(taxa_names(physeq.filtered) %in% longbranch$ID), physeq.filtered)
# OR
# physeq.filtered <- subset_taxa( physeq.filtered, !(Class %in% c("Unclassified","Cyanobacteriia")))
# physeq.filtered
```

How's the tree look now?

```{r}
#p <- plot_tree(physeq.filtered,
#          nodelabf = nodeplotblank,
#          color="Sample",
#          ladderize = "left",
#          method = "treeonly") +
#  ggtitle(paste0("Original tree: ", ntaxa(physeq.filtered), " taxa")) +
#  theme(plot.title = element_text(size = 10))

#ggplotly(p)
```


Agglomeration is based on the cophenetic distance, the pairwise distances between tips on the tree. These are pretty short; let's see what that distribution looks like

```{r PrevalenceFiltering-15}
cp_phylo <- cophenetic.phylo(phy_tree(physeq.filtered))

hist(cp_phylo, 
     breaks = 100, 
     main = "Pairwise distance between tips", 
     xlab = "Distance between tips")
    # xlim = c(-0.5, 3))

cutoff <- c(seq(0.025, 0.175, 0.025), 0.2, 0.3, 0.5, 0.75, 1, 2)
abline(v=cutoff, col = "red")
text(cutoff, max(hist(cp_phylo, 100, plot=FALSE)$counts), labels = cutoff, pos = 2, srt = 90, cex = .5 )
```

The red lines are some arbitrary test cutoffs. There's no clear shoulder at which to make a cutoff, but 0.2 or less look reasonable.

Let's replot in log scale.  

```{r PrevalenceFiltering-16}
hist(log(cp_phylo), 
     breaks = 200, 
     main = "Pairwise distance between tips", 
     xlab = "Distance between tips (log)",
     xlim = c(-5, 2))

abline(v=log(cutoff), col = "red")
text(log(cutoff), max(hist(log(cp_phylo), 100, plot=FALSE)$counts), labels = cutoff, pos = 2, srt = 90, cex = .5 )
```

0.2 seems pretty reasonable since its at the end of the first small peak.

```{r PrevalenceFiltering-17}
# Use the cutoffs listed above

# this takes some time to run :).  There is a speedyseq package with a faster tip_glom implementation, might be worth checking
pseqs <- lapply(cutoff[1:10], function(x) {tip_glom(physeq.filtered, h = x)})

names(pseqs) <- cutoff[1:10]
```

Note there is a `phyloseq` instance with no tree now. Let's only plot the ones that have a tree. 

```{r PrevalenceFiltering-18}
# In order to screen for instances with a tree we need to use tryCatch as checking the tree slot with phy_tree will error if it is NULL)

pseqs.final <- pseqs[sapply(pseqs, function(x) {
  !is.null( tryCatch({phy_tree(x)}, error = function(cond) { return(NULL) }) )
  }, simplify = TRUE)]

plots <- sapply(names(pseqs.final), function(x) {
  plot_tree(pseqs.final[[x]], 
          nodelabf = nodeplotblank,
          ladderize = "left", 
          method = "treeonly") + 
  ggtitle(paste0("Height:",x, ", ", ntaxa(pseqs.final[[x]]), " taxa")) + 
    theme(plot.title = element_text(size = 10))
  }, simplify = FALSE
  )

grid.arrange(grobs = prepend(plots, list(Original = p)),
             nrow = 3)

```
PacBio: Even at the lowest cutoff (0.025), more than half the taxa are removed. Far too much is removed with the 0.5 and 0.75, so those are definitely out of the running. Let's get a closer look at the other cutoffs by changing the number below.

Shoreline: At each cutoff, about 100 are lost, until going from 0.075 to 0.1. Let's get a closer look at the other cutoffs by changing the number below. Last time I went with a cutoff of 0.05, so I should try either 0.075 or 0.1 this time.

```{r PrevalenceFiltering-19}
p <- plot_tree(pseqs.final[['0.15']],
          label.tips = "Genus",
          ladderize = "left",
          justify = "left",
          color = 'Location')
p
```

Here's a closer look.
```{r PrevalenceFiltering-20}
#ggplotly(p)
ggplotly(plot_tree(pseqs.final[['0.1']],
          # nodelabf = nodeplotblank,
          ladderize = "left",
          method = "treeonly"))
```

Both 0.15 and 0.2 seem reasonable when looking at the tips. Could even create a 0.175 cutoff to compromise between the two, however, I think I'll stick with 0.2 for now since it does nicely sit between the two logged peaks.

```{r PrevalenceFiltering-21}
physeq.check <- pseqs.final[['0.2']]

p <- ggplot(data = data.frame(
    SampleLoss = sample_sums(physeq.check) / sample_sums(physeq.filtered),
    Names = factor(sample_names(physeq.check), ordered = TRUE, levels = sample_names(physeq.check)),
    Group = factor(sample_data(physeq.check)$Location, ordered = TRUE)
), aes(y = SampleLoss, x = Names, fill = Group))
p <- p + geom_bar(stat = 'identity' )
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p
```

PacBio: I tested out 0.2, 0.15, and 0.1 to see what their downstream results were. And 0.1 seemed to have the best balance of balancing saving taxonomy and sequences. As you go down in cutoff size you tend to retain more taxa but lose more sequences. So you can see more taxonomic resolution later, but you may have less power to detect significant differences downstream.
Shoreline: I had gone with 0.1

```{r PrevalenceFiltering-22}
physeq.glom <- pseqs.final[['0.1']]
```



## Taxa agglomeration

Phyloseq's tax_glom method is analagous to it's tip_glom method, but it uses categorical data instead of tree data to agglomerate.

What is the effect of taxonomic agglomeration per rank if we were to allow for the removal of NA's? If we don't allow NA's to be removed, then we lose nothing. They just get grouped to the next available ranking. I would not recommend removing NA's as it could skew beta diversity measurements downstream, but it is interesting to see how many sequences are ranked to each level.

What is the effect of taxonomic agglomeration per rank? Let's do a quick run through on the samples; ranks that are not assigned are removed by default, so let's see what happens.

<!-- ```{r PrevalenceFiltering-23 } -->
<!-- taxglom_per_rank = function(physeq.filtered, rank = "Species") { -->
<!--   # TODO: add sanity check -->
<!--   glommedPhyseq <- tax_glom(physeq.filtered, taxrank = rank, NArm = TRUE) -->
<!--   p <- ggplot(data = data.frame( -->
<!--       SampleLoss = sample_sums(glommedPhyseq) / sample_sums(physeq.filtered), -->
<!--       Names = factor(sample_data(glommedPhyseq)$SampleID, -->
<!--                      ordered = TRUE, -->
<!--                      levels = sample_data(glommedPhyseq)$SampleID), -->
<!--       Group = factor(sample_data(glommedPhyseq)$Location, ordered = TRUE) -->
<!--   ), aes(y = SampleLoss, x = Names, fill = Group)) + -->
<!--     geom_bar(stat = 'identity' ) + -->
<!--     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + -->
<!--     ggtitle(paste0("Rank: ", rank)) -->
<!--   return(p) -->
<!-- } -->
<!-- ``` -->


<!-- ```{r PrevalenceFiltering-24 } -->
<!-- tmp <- as(tax_table(physeq.filtered), 'matrix') -->
<!-- tmp[tmp == 'Unclassified'] = NA -->
<!-- tax_table(physeq.filtered) <- tmp -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ranks <- c("Species", "Genus", "Family", "Order") -->

<!-- plots <- lapply(ranks, function(x) { -->
<!--   p <- taxglom_per_rank(physeq.filtered, rank = x) -->
<!--   p + theme(legend.position = "none") + expand_limits(y = c(0, 1)) -->
<!-- }) -->

<!-- grid.arrange(grobs = plots) -->
<!-- ``` -->

<!-- Quite a bit lost with species, but still pretty decent for Genus. We know that Genus will be good for plotting downstream. -->

<!-- Are there rows in there with 'NA'? -->

<!-- ```{r PrevalenceFiltering-25 } -->
<!-- #apply(tax_table(physeq.filtered), 2, function(x) sum(x != "Unclassified")) -->
<!-- apply(tax_table(physeq.filtered), 2, function(x) sum(is.na(x))) -->
<!-- ``` -->


<!-- ```{r} -->
<!--  physeq.taxa.glom <- tax_glom(physeq.filtered, taxrank = 'Species', NArm = FALSE) -->
<!--  physeq.taxa.glom -->
<!-- ``` -->

526 taxa leftover. Closer to the 0.05 tip agglomeration cutoff originally chosen in 2022. -->

UPDATE: I did test the taxa agglomeration method over the tip agglomeration method, and it produced only 36 taxa after prevalence filtering. I am commenting out the code above so that it doesn't get accidentally used.




## Features and Prevalence tables

For the filtering, let's assign the original filtered data to a temp variable prior to prevalence filtering.

```{r PrevalenceFiltering-26 }
physeq0 <- physeq.glom
physeq0
```

Suggested based on the Callahan dada2 workflow (F1000Research, 2017).  This is a bit of data exploration to see how many features are present per taxa.

```{r PrevalenceFiltering-27 }
table(tax_table(physeq0)[,"Phylum"], exclude = NULL)
```
  Acidobacteriota  Actinobacteriota      Bacteroidota  Bdellovibrionota  Campylobacterota       Chloroflexi     Cyanobacteria 
               12                92                31                 2                 2                 9                 1 
     Dependentiae  Desulfobacterota        Firmicutes    Fusobacteriota   Gemmatimonadota Methylomirabilota       Myxococcota 
                1                 3               118                10                 2                 1                 5 
     Nitrospirota   Patescibacteria    Proteobacteria Verrucomicrobiota 
                1                 1               101                 7  
                
There are number with low features (1-2 OTUs). But I don't see any unclassified present, so I'll skip the next step.

```{r PrevalenceFiltering-28 }
#physeq0 <- subset_taxa(physeq0, !is.na(Phylum) & !Phylum %in% c("", "uncharacterized"))
#physeq0
```


Now, let's get an idea how many taxa in the samples have an ASV count greater than 1.  We can make this more or less strict as needed.

```{r PrevalenceFiltering-29 }
# What is this doing?  It calculates a vector with the count being the # samples with a count > 0.

# Note: make sure you are using *raw counts* here; if you use proportional
# counts make sure to adjust the function appropriately
prevdf <- apply(otu_table(physeq0),  # counts
               # use row or column depending on the data
               MARGIN = ifelse(taxa_are_rows(physeq0), yes = 1, no = 2), 
               # how many times the counts in the samples are greater than 0
               FUN = function(x){sum(x > 0)}  
               )
prevdf <- data.frame(Prevalence =  prevdf, # num samples counts are > 0
                     TotalAbundance = taxa_sums(physeq0), # total abundance
                     tax_table(physeq0)) # tax ID and ranks
```

Here is a quick summary of the prevalence results.  These are performed per ASV but summarized at the Phylum rank, with the 

```{r PrevalenceFiltering-30 }
# a quick high level summary at the Phylum rank.
tmp <- plyr::ddply(prevdf, "Phylum", function(df1) { cbind(mean(df1$Prevalence), sum(df1$Prevalence)) })
colnames(tmp) <- c("Phylum", "mean", "sum")
knitr::kable(tmp)
```

We can plot these out to get more resolution.  


The standard prevalence threshold is 0.05 (5%), but this only happens to be 1 sample. I could move this to either 2 or 3 samples by setting it to 0.075 or 0.1, respectively. Since all the groups have at least 3 samples, this would not be unreasonable, however, since the samples differ in library size so much, I am thinking that 2 samples might be better so as not to lose too much diversity.

```{r PrevalenceFiltering-31}
pthresh <- 0.075
```

This is around `r round(pthresh * nsamples(physeq0))` samples.  We can modify this setting, but we'll leave as is for now.  We may want to modify this to not reflect the specific group but the treatments (e.g. ensure we're not losing any taxa based on the treatment condition)

This plot shows the fraction of samples vs the total abundance for that, which helps give some idea on what to retain.

```{r PrevalenceFiltering-32 }
ggplot(prevdf,
       aes(TotalAbundance, Prevalence / nsamples(physeq0), color = Phylum)) +
  geom_hline(yintercept = pthresh, alpha = 0.5, linetype = 2) +
  geom_point(size = 2, alpha = 0.4) +
  scale_x_log10() +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position = "none")
```

Either cutoff of 2 or 3 samples doesn't seem to unreasonable. But sticking to 2 samples for now.

```{r PrevalenceFiltering-33 }
prevThreshold <- pthresh * nsamples(physeq.glom)

keepTaxa <- rownames(prevdf)[(prevdf$Prevalence >= prevThreshold)]
physeq.prev <- prune_taxa(keepTaxa, physeq.glom)
physeq.prev
```
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 94 taxa and 28 samples ]
sample_data() Sample Data:       [ 28 samples by 14 sample variables ]
tax_table()   Taxonomy Table:    [ 94 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 94 tips and 93 internal nodes ]
refseq()      DNAStringSet:      [ 94 reference sequences ]

This reduces it down considerably!  From 399 taxa to 94. At the other tip glom cutoffs it was 65 or 60 taxa at the end. 

```{r PrevalenceFiltering-34 }
p <- ggplot(data = data.frame(
    SampleLoss = sample_sums(physeq.prev) / sample_sums(physeq.glom),
    Names = factor(sample_names(physeq.prev), ordered = TRUE, levels = sample_names(physeq.prev)),
    Group = factor(sample_data(physeq.prev)$Location, ordered = TRUE)
), aes(y = SampleLoss, x = Names, fill = Group))
p <- p + geom_bar(stat = 'identity' )
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p
```

There is quite a bit of count/sequence loss in about 10-11 samples

There's a balance of retaining taxa versus retaining sequences per sample. Lowering the tip agglomeration threshold down to 0.1 created 94 taxa, but then resulted in a lot more sequence loss than 0.2 and 0.15. I ended up testing the beta diversity results on both 0.2 and 0.15 to confirm that this was the right choice, and 0.1 is the most taxonomically conservative option while still maintaining power for statistical tests later.

# For testing

Let's also save some of the other prevalence threshold cutoffs

```{r PrevalenceFiltering-33 }
# Threshold of 0.15
prevdf <- apply(otu_table(pseqs.final[['0.15']]),  # counts
               # use row or column depending on the data
               MARGIN = ifelse(taxa_are_rows(pseqs.final[['0.15']]), yes = 1, no = 2), 
               # how many times the counts in the samples are greater than 0
               FUN = function(x){sum(x > 0)}  
               )
prevdf <- data.frame(Prevalence =  prevdf, # num samples counts are > 0
                     TotalAbundance = taxa_sums(pseqs.final[['0.15']]), # total abundance
                     tax_table(pseqs.final[['0.15']])) # tax ID and ranks

prevThreshold <- pthresh * nsamples(pseqs.final[['0.15']])
keepTaxa <- rownames(prevdf)[(prevdf$Prevalence >= prevThreshold)]
physeq.prev.15 <- prune_taxa(keepTaxa, pseqs.final[['0.15']])
physeq.prev.15
```


# Save

We'll save at this stage, and then reload the data for diversity analysis and differential abundance.

```{r PrevalenceFiltering-35 }
# Save
if (!file.exists('./results/PrevalenceFiltering/')){
    dir.create(file.path('./results/PrevalenceFiltering/'), recursive = TRUE)
}
saveRDS(physeq.prev, file = "results/PrevalenceFiltering/phyloseq.prevfiltered.RDS")
#saveRDS(physeq.prev.15, file = "results/PrevalenceFiltering/phyloseq.prevfiltered.15.RDS")
```



# Session information

```{r PrevalenceFiltering-36 }
sessionInfo()
```

R version 4.3.0 Patched (2023-06-03 r84490)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.5.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] phyloseq.extended_0.0.0.9000 plotly_4.10.2                picante_1.8.2               
 [4] dada2_1.28.0                 Rcpp_1.0.10                  readxl_1.4.2                
 [7] gridExtra_2.3                reshape2_1.4.4               scales_1.2.1                
[10] exactRankTests_0.8-35        Biostrings_2.68.1            GenomeInfoDb_1.36.0         
[13] XVector_0.40.0               IRanges_2.34.0               S4Vectors_0.38.1            
[16] ranacapa_0.1.0               microbiome_1.22.0            plotly.microbiome_0.0.9003  
[19] devtools_2.4.5               usethis_2.2.0                lsmeans_2.30-0              
[22] emmeans_1.8.8                nlme_3.1-162                 lmerTest_3.1-3              
[25] lme4_1.1-33                  WGCNA_1.72-1                 fastcluster_1.2.3           
[28] dynamicTreeCut_1.63-1        mixOmics_6.24.0              MASS_7.3-60                 
[31] ggpubr_0.6.0                 vegan_2.6-4                  lattice_0.21-8              
[34] permute_0.9-7                ggdendro_0.1.23              biomformat_1.28.0           
[37] ape_5.7-1                    metagenomeSeq_1.42.0         RColorBrewer_1.1-3          
[40] glmnet_4.1-8                 Matrix_1.5-4.1               limma_3.56.2                
[43] Biobase_2.60.0               BiocGenerics_0.46.0          phyloseq_1.44.0             
[46] lubridate_1.9.2              forcats_1.0.0                stringr_1.5.0               
[49] dplyr_1.1.2                  purrr_1.0.1                  readr_2.1.4                 
[52] tidyr_1.3.0                  tibble_3.2.1                 ggplot2_3.4.2               
[55] tidyverse_2.0.0              knitr_1.43                  

loaded via a namespace (and not attached):
  [1] fs_1.6.2                    matrixStats_1.0.0           bitops_1.0-7               
  [4] httr_1.4.6                  doParallel_1.0.17           numDeriv_2016.8-1.1        
  [7] profvis_0.3.8               tools_4.3.0                 backports_1.4.1            
 [10] utf8_1.2.3                  R6_2.5.1                    lazyeval_0.2.2             
 [13] mgcv_1.8-42                 rhdf5filters_1.12.1         urlchecker_1.0.1           
 [16] withr_2.5.0                 prettyunits_1.1.1           preprocessCore_1.62.1      
 [19] textshaping_0.3.6           cli_3.6.1                   labeling_0.4.2             
 [22] mvtnorm_1.2-3               dittoSeq_1.12.0             ggridges_0.5.4             
 [25] systemfonts_1.0.4           Rsamtools_2.16.0            foreign_0.8-84             
 [28] sessioninfo_1.2.2           rstudioapi_0.14             impute_1.74.1              
 [31] RSQLite_2.3.1               generics_0.1.3              shape_1.4.6                
 [34] hwriter_1.3.2.1             crosstalk_1.2.0             vroom_1.6.3                
 [37] gtools_3.9.4                car_3.1-2                   GO.db_3.17.0               
 [40] interp_1.1-4                fansi_1.0.4                 abind_1.4-5                
 [43] lifecycle_1.0.3             yaml_2.3.7                  carData_3.0-5              
 [46] SummarizedExperiment_1.30.1 Rtsne_0.16                  gplots_3.1.3               
 [49] rhdf5_2.44.0                grid_4.3.0                  blob_1.2.4                 
 [52] promises_1.2.0.1            crayon_1.5.2                miniUI_0.1.1.1             
 [55] cowplot_1.1.1               KEGGREST_1.40.0             pillar_1.9.0               
 [58] GenomicRanges_1.52.0        boot_1.3-28.1               estimability_1.4.1         
 [61] corpcor_1.6.10              codetools_0.2-19            glue_1.6.2                 
 [64] ShortRead_1.58.0            data.table_1.14.8           remotes_2.4.2              
 [67] vctrs_0.6.2                 png_0.1-8                   testthat_3.1.8             
 [70] cellranger_1.1.0            gtable_0.3.3                cachem_1.0.8               
 [73] xfun_0.39                   S4Arrays_1.0.4              mime_0.12                  
 [76] survival_3.5-5              pheatmap_1.0.12             SingleCellExperiment_1.22.0
 [79] iterators_1.0.14            ellipsis_0.3.2              bit64_4.0.5                
 [82] rprojroot_2.0.3             KernSmooth_2.23-21          rpart_4.1.19               
 [85] colorspace_2.1-0            DBI_1.1.3                   Hmisc_5.1-0                
 [88] nnet_7.3-19                 ade4_1.7-22                 NADA_1.6-1.1               
 [91] tidyselect_1.2.0            processx_3.8.1              bit_4.0.5                  
 [94] compiler_4.3.0              htmlTable_2.4.1             desc_1.4.2                 
 [97] DelayedArray_0.26.3         checkmate_2.2.0             caTools_1.18.2             
[100] callr_3.7.3                 digest_0.6.31               minqa_1.2.5                
[103] rmarkdown_2.22              jpeg_0.1-10                 htmltools_0.5.5            
[106] pkgconfig_2.0.3             base64enc_0.1-3             MatrixGenerics_1.12.0      
[109] fastmap_1.1.1               rlang_1.1.1                 htmlwidgets_1.6.2          
[112] shiny_1.7.4                 zCompositions_1.4.0-1       farver_2.1.1               
[115] jsonlite_1.8.5              BiocParallel_1.34.2         RCurl_1.98-1.12            
[118] magrittr_2.0.3              Formula_1.2-5               GenomeInfoDbData_1.2.10    
[121] Rhdf5lib_1.22.0             munsell_0.5.0               viridis_0.6.3              
[124] stringi_1.7.12              brio_1.1.3                  zlibbioc_1.46.0            
[127] plyr_1.8.8                  pkgbuild_1.4.0              ggrepel_0.9.3              
[130] deldir_1.0-9                splines_4.3.0               multtest_2.56.0            
[133] hms_1.1.3                   locfit_1.5-9.8              ps_1.7.5                   
[136] igraph_1.4.3                ggsignif_0.6.4              Wrench_1.18.0              
[139] pkgload_1.3.2               evaluate_0.21               latticeExtra_0.6-30        
[142] RcppParallel_5.1.7          nloptr_2.0.3                tzdb_0.4.0                 
[145] foreach_1.5.2               httpuv_1.6.11               broom_1.0.4                
[148] xtable_1.8-4                RSpectra_0.16-1             rstatix_0.7.2              
[151] later_1.3.1                 ragg_1.2.5                  viridisLite_0.4.2          
[154] truncnorm_1.0-9             rARPACK_0.11-0              GenomicAlignments_1.36.0   
[157] memoise_2.0.1               AnnotationDbi_1.62.1        ellipse_0.4.5              
[160] cluster_2.1.4               timechange_0.2.0 